ambry-store/src/main/java/com/github/ambry/store/PersistentIndex.java

```diff
@@ -1141,28 +1175,33 @@ IndexValue markAsUndeleted(StoreKey id, FileSpan fileSpan, MessageInfo info, lon
    * @throws StoreException
    */
   BlobReadOptions getBlobReadInfo(StoreKey id, EnumSet<StoreGetOptions> getOptions) throws StoreException {
-    ConcurrentSkipListMap<Offset, IndexSegment> indexSegments = validIndexSegments;
-    IndexValue value = findKey(id);
-    BlobReadOptions readOptions;
-    if (value == null) {
-      throw new StoreException("Id " + id + " not present in index " + dataDir, StoreErrorCodes.ID_Not_Found);
-    } else if (value.isDelete()) {
-      if (!getOptions.contains(StoreGetOptions.Store_Include_Deleted)) {
-        throw new StoreException("Id " + id + " has been deleted in index " + dataDir, StoreErrorCodes.ID_Deleted);
+    rwLock.readLock().lock();
```

Comment:
Getting the lock and releasing it upon the conclusion of this function is not sufficient. Blobstore::get function is accessing the log segment pointed by the BlobReadOptions returned by this function. In other words, when the Blobstore::get attempts to access the log segment, the log segment could be already deleted by the compaction. For that reason, this lock needs to be owned by the Blobstore and acquired within the scope of Blobstore::get operation, i.e. acquiring the lock should guarantee that index and log accesses will be coherent.

Reply:
It's actually fine in this case since the compaction thread would wait for the log segment's refCount becoming 0. The Lock would make sure that these three steps 1. get index value from persistent index 2 getting log segment from index value 3 increase the refCount of the log segment   would become atomic.
That's why we don't have to lock the BlobStore.get, and only on the PersistentIndex.getBlobReadInfo. When getBlobReadInfo function returns a BlobReadOptions, it already increased the refCount of the log segment. We know the refCount is not 0, so the compaction thread will be waiting for the refCount to become 0 before it remove the log segment from the map and from the disk.